semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/cfgdominator/CfgDominatorFrontier.scala

```diff
@@ -2,39 +2,37 @@ package io.shiftleft.semanticcpg.passes.cfgdominator
 
 import scala.collection.mutable
 
+/**
+  * Calculates the dominator frontier for a set of CFG nodes.
+  * The returned multimap associates the frontier nodes to each CFG node.
+  *
+  * The used algorithm is from: "A Simple, Fast Dominance Algorithm" from
+  * "Keith D. Cooper, Timothy J. Harvey, and Ken Kennedy".
+  */
 class CfgDominatorFrontier[NodeType](cfgAdapter: CfgAdapter[NodeType], domTreeAdapter: DomTreeAdapter[NodeType]) {
 
-  /**
-    * Calculates a the dominator frontier for a set of CFG nodes.
-    * The returned multimap associates the frontier nodes to each CFG node.
-    *
-    * The used algorithm is from: "A Simple, Fast Dominance Algorithm" from
-    * "Keith D. Cooper, Timothy J. Harvey, and Ken Kennedy".
-    */
-  def calculate(allCfgNodes: Seq[NodeType]): mutable.MultiDict[NodeType, NodeType] = {
-    val domFrontier = mutable.MultiDict.empty[NodeType, NodeType]
-
-    allCfgNodes.foreach { joinCandiate =>
-      val predecessors = cfgAdapter.predecessors(joinCandiate).iterator.to(Seq)
-
-      if (predecessors.size > 1) {
-        val joinNode = joinCandiate
+  private def doms(x: NodeType): Option[NodeType] = domTreeAdapter.immediateDominator(x)
+  private def pred(x: NodeType): Seq[NodeType] = cfgAdapter.predecessors(x).iterator.to(Seq)
+  private def onlyJoinNodes(x: NodeType): Option[(NodeType, Seq[NodeType])] =
+    Some(pred(x)).filter(_.size > 1).map(p => (x, p))
+  private def withIDom(x: NodeType, preds: Seq[NodeType]) =
+    doms(x).map(i => (x, preds, i))
 
-        domTreeAdapter.immediateDominator(joinNode) match {
-          case Some(immediateJoinNodeDom) =>
-            predecessors.foreach { predecessor =>
-              var currentPred = Option(predecessor)
-
-              while (currentPred.isDefined && currentPred.get != immediateJoinNodeDom) {
-                domFrontier.addOne(currentPred.get, joinNode)
-                currentPred = domTreeAdapter.immediateDominator(currentPred.get)
-              }
+  def calculate(cfgNodes: Seq[NodeType]): mutable.MultiDict[NodeType, NodeType] = {
+    val domFrontier = mutable.MultiDict.empty[NodeType, NodeType]
+    cfgNodes
```

Comment:
this would look much nicer in a for comprehension - let me know if you're interested and I'll port it over

Reply:
Applied the first three changes. Can you show me what this looks like with a for-comprehension? Since I'm pointing to this code as part of the teaching material on implementing a dominator frontier algorithm, I'm always wondering which features that are specific to Scala I should make use of and which ones I should rather leave out.

Reply:
Looks great, I'll take it ;)